compiler/compiler.go

```diff
@@ -84,6 +84,9 @@ func substituteTypes(subst *semantic.Substitution, inferredType, actualType sema
 	// which isn't represented in type inference.
 	if actualType.Nature() == semantic.Invalid {
 		return nil
+	} else if actualType.Nature() == semantic.Dynamic {
+		// Dynamic values are as reduced as they are going to get.
+		return nil
```

Comment:
Seems like ideally this substitution would happen in rust and flow back through here.

I'm not yet sure how to make that happen.